File: app/src/main/java/com/polestar/charging/ui/cluster/view/CreateMarkerTask.java
package com.polestar.charging.ui.cluster.view;

import com.amap.api.maps.model.LatLng;
import com.amap.api.maps.model.Marker;
import com.amap.api.maps.model.MarkerOptions;
import com.polestar.charging.ui.cluster.base.Cluster;
import com.polestar.charging.ui.cluster.base.ClusterItem;
import com.polestar.charging.ui.cluster.view.renderer.MarkerWithPosition;

import java.util.Set;

/**
 * Creates markerWithPosition(s) for a particular cluster, animating it if necessary.
 */
public class CreateMarkerTask<T extends ClusterItem> {
    private final Cluster<T> cluster;
    private final Set<MarkerWithPosition> newMarkers;
    private final LatLng animateFrom;
    DefaultClusterRenderer renderer;

    public CreateMarkerTask(Cluster<T> c, Set<MarkerWithPosition> markersAdded, LatLng animateFrom,    DefaultClusterRenderer renderer) {
        this.cluster = c;
        this.newMarkers = markersAdded;
        this.animateFrom = animateFrom;
        this.renderer = renderer;
    }

    void perform(MarkerModifier markerModifier) {
        // Don't show small clusters. Render the markers inside, instead.
        if (!shouldRenderAsCluster(cluster)) {
            for (T item : cluster.getItems()) {
                Marker marker = renderer.mMarkerCache.get(item);
                MarkerWithPosition markerWithPosition;
                if (marker == null) {
                    MarkerOptions markerOptions = new MarkerOptions();
                    if (animateFrom != null) {
                        markerOptions.position(animateFrom);
                    } else {
                        markerOptions.position(item.getPosition());
                    }
                    renderer.mMarkerManager.onBeforeClusterItemRendered(
                            item.getTitle(),
                            item.getSnippet(),
                            markerOptions
                    );
                    marker = renderer.mSingleMarkers.addMarker(markerOptions);
                    markerWithPosition = new MarkerWithPosition(marker);
                    renderer.mMarkerCache.put(item, marker);
                    if (animateFrom != null) {
                        markerModifier.animate(markerWithPosition, animateFrom, item.getPosition());
                    }
                } else {
                    markerWithPosition = new MarkerWithPosition(marker);
                    renderer.mMarkerManager.onClusterItemUpdated(item.getTitle(), item.getSnippet(), item.getPosition(), marker);
                }
                newMarkers.add(markerWithPosition);
            }
            return;
        }

        Marker marker = renderer.mClusterMarkerCache.get(cluster);
        MarkerWithPosition markerWithPosition;
        if (marker == null) {
            MarkerOptions markerOptions = new MarkerOptions().
                    position(animateFrom == null ? cluster.getPosition() : animateFrom);
            markerOptions.icon(renderer.mIconGenerator.getDescriptorForCluster(cluster.getSize()));
            marker = renderer.mClusterMarkers.addMarker(markerOptions);
            renderer.mClusterMarkerCache.put(cluster, marker);
            markerWithPosition = new MarkerWithPosition(marker);
            if (animateFrom != null) {
                markerModifier.animate(markerWithPosition, animateFrom, cluster.getPosition());
            }
        } else {
            markerWithPosition = new MarkerWithPosition(marker);
            marker.setIcon(renderer.mIconGenerator.getDescriptorForCluster(cluster.getSize()));
        }
        newMarkers.add(markerWithPosition);
    }

    boolean shouldRenderAsCluster(Cluster<T> cluster) {
        return cluster.getSize() >= DefaultClusterRenderer.MIN_CLUSTER_SIZE;
    }
}

File: app/src/main/java/com/polestar/charging/ui/cluster/CreateMarkerTask.kt
package com.polestar.charging.ui.cluster

import com.amap.api.maps.model.LatLng
import com.amap.api.maps.model.Marker
import com.amap.api.maps.model.MarkerOptions
import com.polestar.charging.ui.cluster.base.Cluster
import com.polestar.charging.ui.cluster.base.ClusterItem

class CreateMarkerTask<T : ClusterItem>(
    val cluster: Cluster<T>,
    val newMarkers: MutableSet<MarkerWithPosition>,
    val animateFrom: LatLng?,
    val zoomCallback: RenderTask.OnZoomChange<T>
) {
    fun perform(markerModifier: MarkerModifier<T>) {
        if (cluster.size > 4) {
            cluster.items?.forEach { item ->
                zoomCallback.getMarkerCacheItem(item)?.let { marker ->
                    item.title?.let {
                        marker.title = it
                    }
                    item.snippet?.let {
                        marker.snippet = it
                    }
                    MarkerWithPosition(marker)
                } ?: run {
                    val options = MarkerOptions()
                        .position(animateFrom ?: item.position)
                    onBeforeClusterItemRendered(item, options)
                    val marker: Marker = zoomCallback.addMarker(options)
                    val markerWithPosition = MarkerWithPosition(marker)
                    zoomCallback.markerCachePut(item, marker)
                    if (animateFrom != null) {
                        markerModifier.animate(markerWithPosition, animateFrom, item.position);
                    }
                    markerWithPosition
                }.let {
                    newMarkers.add(it);
                }
            }
        } else {
            var marker: Marker? = zoomCallback.getClusterMarkerCache(cluster)
            marker?.let {
                it.title = "C:" + cluster.size
                val markerWithPosition = MarkerWithPosition(it);
                markerWithPosition
            } ?: run {
                val markerOptions = MarkerOptions()
                    .position(animateFrom ?: cluster.position)
                onBeforeClusterRendered(cluster, markerOptions)

                marker = zoomCallback.addClusterMarker(markerOptions)
                zoomCallback.clusterCachePut(cluster, marker!!)
                val markerWithPosition = MarkerWithPosition(marker!!)

                if (animateFrom != null) {
                    markerModifier.animate(markerWithPosition, animateFrom, cluster.position!!)
                }
                markerWithPosition
            }.let {
                newMarkers.add(it);
            }
        }
    }

    fun onBeforeClusterItemRendered(item: T, markerOptions: MarkerOptions) {
        item.title?.let {
            markerOptions.title(it)
        }
        item.snippet?.let {
            markerOptions.snippet(it)
        }
    }

    fun onBeforeClusterRendered(cluster: Cluster<T>, markerOptions: MarkerOptions) {
        markerOptions.title("C:" + cluster.size)
    }
}

